### PR TITLE
feat(core): add precision-recall AUC metric

### LINKED ISSUE #1509: Average precision metric for binary classification
**Is your feature request related to a problem? Please describe.**
There is ROC AUC in `ComputeModelStatistics`, but at the same time Average Precision (`areaUnderPR`) is absent.

**Describe the solution you'd like**
It will be awesome to add it as it is very useful for many binary classification tasks. 

**Additional context**
None.

AB#1789611

### PR DESCRIPTION
**Pull request by ranadeepsingh**
## Summary

- expose binary precision-recall AUC as `areaUnderPR` through `ComputeModelStatistics`
- include it in binary `all` / `classification` results and AutoML metric selection
- document and test `areaUnderROC` as an alias for the unchanged `AUC` ROC result
- harden scoped release-compat prerequisite parsing for CRLF and surrounding whitespace

## Metric semantics

`areaUnderPR` delegates directly to Spark's `BinaryClassificationMetrics.areaUnderPR()`. This is trapezoidal area under the precision-recall curve and intentionally differs from step-wise average precision, so this change does not add an `averagePrecision` alias.

Existing `AUC` behavior remains ROC AUC. Binary aggregate output retains the existing metric order and appends `areaUnderPR` after `AUC`.

## Compatibility

- no existing public metric names, parameters, or serialized parameter shapes are removed or renamed
- `AUC` values and `rocCurve` behavior remain unchanged
- `transformSchema` advertises binary-only aggregate columns only when SynapseML categorical metadata proves at most two label levels
- multiclass and unknown-cardinality inputs preserve the legacy common metric-only schema contract
- `evaluation_type`, `confusion_matrix`, and multiclass detail columns remain runtime-only for source compatibility
- aggregate ROC and PR calculations reuse and unpersist one `BinaryClassificationMetrics` instance

## Tests

The focused imbalanced fixture has 2 positives among 100 uniquely ranked examples and asserts:

- ROC AUC: `187 / 196 = 0.9540816326530612`
- PR AUC: `251 / 440 = 0.5704545454545454`
- conventional AP would be `13 / 22`, confirming the semantics are not mislabeled

Additional coverage validates binary and multiclass aggregate schemas/runtime ordering, unknown-cardinality compatibility, multiclass rejection, AutoML ordering/validation/save behavior, and normalized/CRLF/trailing-whitespace prerequisite parsing.

## Validation

- rebased onto `microsoft/master` at `55e4c2593d`
- JDK 11 SynapseML wrapper: focused schema tests **4/4 passed**; remaining targeted metric/AutoML suites **8/8 passed**
- `core/compile` and `core/test:compile`: passed
- repository `scalastyle` and `test:scalastyle`: 0 errors / warnings
- Black 22.3: 189 files unchanged
- Linux CI replay pytest: **22 passed**
- Spark 4.1 release replay, including scoped prerequisite `6938c472...`, followed by `core/test:compile` on JDK 17: passed
- no public Param or generated wrapper shape changed in the review-fix commit; prior codegen/wrapper validation remains applicable
- the prior causal job failure occurred before tests started due an Azure DevOps TLS certificate hostname mismatch; it produced no test results and is infrastructure-related

Fixes #1509

### CODE REVIEW

File: core/src/test/scala/com/microsoft/azure/synapse/ml/train/VerifyComputeModelStatistics.scala [hunk 1]
```diff
@@ -41,6 +41,110 @@ class VerifyComputeModelStatistics extends TransformerFuzzing[ComputeModelStatis
...
+    val result = evaluator.transform(rankedBinaryDataset)
+    val areaUnderPR = result.first().getAs[Double](MetricConstants.AreaUnderPRColumnName)
+
+    assert(result.columns.last === MetricConstants.AreaUnderPRColumnName)
+    assert(result.columns.contains(MetricConstants.AreaUnderPRColumnName))
+    assert(evaluator.transformSchema(rankedBinaryDataset.schema) ===
+      StructType(Array(StructField(MetricConstants.AreaUnderPRColumnName, DoubleType))))
+    assert(math.abs(areaUnderPR - 251.0 / 440.0) < 1e-12)
+    assert(math.abs(areaUnderPR - 13.0 / 22.0) > 0.01)
+  }
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply core/src/test/scala/com/microsoft/azure/synapse/ml/train/VerifyComputeModelStatistics.scala:96]: The five new PR-AUC/ROC numeric assertions were relaxed from `1e-12` to `1e-8` in commit `6291f1d892877e38e3900e771b61d2bf54bdfa83`, and validation passed: `VerifyComputeModelStatistics` 17/17 and `core/test:scalastyle` 0 errors/0 warnings. Thanks for the review!

File: pipeline.yaml [hunk 2]
```diff
@@ -1028,21 +1028,112 @@ jobs:
...
             esac
+            if [[ "$RAW_LINE" == *$'\t'* ]]; then
+              PREREQUISITE="${RAW_LINE%%$'\t'*}"
+              PREREQUISITE_SCOPE="${RAW_LINE#*$'\t'}"
+              PREREQUISITE_IS_SCOPED=true
+            else
+              PREREQUISITE="$TRIMMED_LINE"
+              PREREQUISITE_SCOPE=""
+              PREREQUISITE_IS_SCOPED=false
+            fi
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply pipeline.yaml]: Fixed in 7857fde6aa. Scoped entries now detect and split tabs from TRIMMED_LINE, so CRLF and surrounding whitespace cannot leak into the SHA or final path. Regression coverage includes normalized LF, CRLF, and surrounding-whitespace configs; the full Linux replay suite passes 22/22.

File: core/src/main/scala/com/microsoft/azure/synapse/ml/train/ComputeModelStatistics.scala [hunk 3]
```diff
@@ -439,31 +485,63 @@ class ComputeModelStatistics(override val uid: String) extends Transformer
...
         if (isDefined(labelCol)) Some(getLabelCol) else None,
         getEvaluationMetric)
-    val columns =
-      if (scoreValueKind == SchemaConstants.ClassificationKind) MetricConstants.ClassificationColumns
-      else if (scoreValueKind == SchemaConstants.RegressionKind) MetricConstants.RegressionColumns
+    val (columns, validMetrics) =
+      if (scoreValueKind == SchemaConstants.ClassificationKind)
+        (MetricConstants.ClassificationColumns, MetricConstants.ClassificationMetrics)
+      else if (scoreValueKind == SchemaConstants.RegressionKind)
+        (MetricConstants.RegressionColumns, MetricConstants.RegressionMetrics)
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply core/src/main/scala/com/microsoft/azure/synapse/ml/train/ComputeModelStatistics.scala:498]: Fixed in 7857fde6aa. transformSchema now returns BinaryClassificationColumns when SynapseML categorical metadata proves <=2 label levels, while multiclass and unknown-cardinality inputs retain the legacy common metric-only schema. Tests assert exact binary and multiclass runtime ordering plus schema compatibility. JDK 11 focused tests pass 4/4, and the release-replayed Spark 4.1 core/test:compile passes on JDK 17.

**Comment by ranadeepsingh** [comment]
Review follow-up is pushed at 7857fde6aa: both live threads are resolved, suppressed documentation feedback is addressed, Linux release-replay tests pass 22/22, JDK 11 core compile/style and targeted suites pass, and the scoped patch compiles against Spark 4.1 on JDK 17. The previous causal failure was infrastructure-only: AzureCLI failed on an msdata.visualstudio.com TLS hostname mismatch before any tests ran, leaving no test results.

File: pipeline.yaml [hunk 4]
```diff
@@ -1029,21 +1029,112 @@ jobs:
...
+                if [ -z "$path" ]; then
+                  echo "##vso[task.logissue type=error]Scoped prerequisite $PREREQUISITE contains an empty path"
+                  exit 1
+                fi
+                case "$path" in
+                  /*|.|./*)
+                    echo "##vso[task.logissue type=error]Scoped prerequisite path must be repo-relative: $path"
+                    exit 1
+                    ;;
+                esac
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply pipeline.yaml:1085]: Fixed in b1514ee449. The Bash validator now rejects empty path components and exact dot components before replay matching: repeated slashes, interior/terminal `.` segments, and trailing slashes all fail with a normalized-path error. Existing absolute/leading-dot and `..` traversal checks remain in place.

File: tools/ci/tests/test_pipeline_yaml.py [hunk 5]
```diff
@@ -379,17 +424,22 @@ def test_release_compat_accepts_github_target_and_uses_one_sbt_process():
...
-    assert all(re.fullmatch(r"[0-9a-fA-F]{40}", line) for line in lines)
-    assert len(lines) == len(set(lines)), "prerequisite commits must be unique"
+    # backports, and the replay script already treats an absent or empty list as a no-op.
+    entries = [line.split("\t") for line in lines]
+    assert all(re.fullmatch(r"[0-9a-fA-F]{40}", fields[0]) for fields in entries)
+    assert all(all(fields[1:]) for fields in entries)
+    for _, *paths in entries:
+        assert all(not path.startswith("/") for path in paths)
+        assert all(".." not in path.split("/") for path in paths)
+    shas = [fields[0] for fields in entries]
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply tools/ci/tests/test_pipeline_yaml.py:458]: Fixed in b1514ee449. The Python config guard now applies the same component rules as Bash, with accepted edge cases (`.hidden`, `...`) and rejected absolute, empty, `.`, `..`, repeated-slash, terminal-dot, and trailing-slash cases. Parameterized tests execute the actual embedded Bash for all newly reported forms. Full Linux replay validation passes 40/40.

**Comment by ranadeepsingh** [comment]
Follow-up b1514ee449 rejects non-normalized scoped prerequisite paths in Bash and mirrors the rule in Python config validation. Actual-script regressions cover dot segments, repeated slashes, trailing slashes, and terminal dot components. Linux replay tests pass 40/40; Black 22.3 passes. Head remains based on upstream master 55e4c2593d71af86dd71feb26dc394826a919497.

File: tools/ci/tests/test_pipeline_yaml.py [hunk 6]
```diff
@@ -70,6 +70,39 @@ def _git(repo, *args):
...
+    assert not status, f"{context} left scratch repo dirty:\n{status}"
+
+
+def _is_normalized_prerequisite_path(path):
+    return (
+        bool(path)
+        and not path.startswith("/")
+        and not path.endswith("/")
+        and all(part not in {"", ".", ".."} for part in path.split("/"))
+    )
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply tools/ci/tests/test_pipeline_yaml.py:106]: Fixed in 901ad8cc10. `_is_normalized_prerequisite_path()` now rejects ASCII edge whitespace and tab/CR/LF contamination while preserving valid internal spaces, `.hidden` segments, and `...` names. Positive/negative helper cases cover leading/trailing spaces, tabs, CR/LF, repeated internal spaces, and existing path normalization rules.

File: pipeline.yaml [hunk 7]
```diff
@@ -1029,21 +1029,116 @@ jobs:
...
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply pipeline.yaml:1085]: Fixed in 901ad8cc10. The parser removes only the terminal CR used by CRLF, preserves the scoped fields for validation, and rejects each extracted path when it differs from its edge-whitespace-trimmed form. Actual-script tests cover clean CRLF, trailing-space+CRLF, leading/trailing spaces in either path field, extra leading/trailing tab fields, and valid paths with internal spaces. Linux replay tests pass 58/58.

**Comment by ranadeepsingh** [comment]
Follow-up 901ad8cc10 rejects leading/trailing whitespace per scoped path field while preserving clean CRLF files, leading whitespace before the SHA, and valid internal spaces. Linux replay validation passes 58/58 and Black 22.3 leaves 189 files unchanged. Head remains based on upstream master 55e4c2593d71af86dd71feb26dc394826a919497.
